src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,461 @@
+                {
+                    firstEmptySubscriptionIndex = subscriptionIndex;
+                }
+            }
+        }
+    }
+
+    // Fail if no empty space
+    if (firstEmptySubscriptionIndex == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    // Now construct subscription state and save
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    backingBuffer.Calloc(MaxSubscriptionSize());
+    ReturnErrorCodeIf(backingBuffer.Get() == nullptr, CHIP_ERROR_NO_MEMORY);
+
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), MaxSubscriptionSize());
+
+    ReturnErrorOnFailure(Save(writer, subscriptionInfo));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+
+    ReturnErrorOnFailure(
+        mStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumption(firstEmptySubscriptionIndex).KeyName(),
+                                  backingBuffer.Get(), static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::Delete(NodeId nodeId, FabricIndex fabricIndex, SubscriptionId subscriptionId)
+{
+    CHIP_ERROR deleteErr = CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+
+    uint16_t count = 0;
+    for (uint16_t subscriptionIndex = 0; subscriptionIndex < CHIP_IM_MAX_NUM_SUBSCRIPTIONS; subscriptionIndex++)
+    {
+        SubscriptionInfo subscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, subscriptionInfo);
+
+        // delete duplicate
+        if (err == CHIP_NO_ERROR)
+        {
+            if ((nodeId == subscriptionInfo.mNodeId) && (fabricIndex == subscriptionInfo.mFabricIndex) &&
+                (subscriptionId == subscriptionInfo.mSubscriptionId))
+            {
+                deleteErr = Delete(subscriptionIndex);
+            }
+            else
+            {
+                count++;
+            }
+        }
+    }
+
+    // if there are no persisted subscriptions, the MaxCount can also be deleted
+    if (count == 0)
+    {
+        DeleteMaxCount();
+    }
+
+    return deleteErr;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::DeleteMaxCount()
+{
+    return mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionMaxCount().KeyName());
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::DeleteAll(FabricIndex fabricIndex)
+{
+    CHIP_ERROR deleteErr = CHIP_NO_ERROR;
+
+    uint16_t count = 0;
+    for (uint16_t subscriptionIndex = 0; subscriptionIndex < CHIP_IM_MAX_NUM_SUBSCRIPTIONS; subscriptionIndex++)
+    {
+        SubscriptionInfo subscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, subscriptionInfo);
+
+        if (err == CHIP_NO_ERROR)
+        {
+            if (fabricIndex == subscriptionInfo.mFabricIndex)
+            {
+                err = Delete(subscriptionIndex);
```

Comment:
Similar questions as above: this makes the last err win, so if a delete fails but subsequent succeed, only the success is kept.

Reply:
Marking resolved .. I see here we check for CHIP_NO_ERROR